format/FlightSql.proto

```diff
@@ -0,0 +1,454 @@
+ *
+ * The returned schema will be:
+ * <
+ *  pk_catalog_name: utf8,
+ *  pk_schema_name: utf8,
+ *  pk_table_name: utf8,
+ *  pk_column_name: utf8,
+ *  fk_catalog_name: utf8,
+ *  fk_schema_name: utf8,
+ *  fk_table_name: utf8,
+ *  fk_column_name: utf8,
+ *  key_sequence: int,
+ *  fk_key_name: utf8,
+ *  pk_key_name: utf8,
+ *  update_rule: uint1,
+ *  delete_rule: uint1
+ * >
+ * The returned data should be ordered by pk_catalog_name, pk_schema_name, pk_table_name, pk_key_name, then key_sequence.
+ * update_rule and delete_rule returns a byte that is equivalent to actions:
+ *    - 0 = CASCADE
+ *    - 1 = RESTRICT
+ *    - 2 = SET NULL
+ *    - 3 = NO ACTION
+ *    - 4 = SET DEFAULT
+ */
+message CommandGetImportedKeys {
+  option (experimental) = true;
+
+  // Specifies the catalog to search for the primary key table.
+  google.protobuf.StringValue catalog = 1;
+
+  // Specifies the schema to search for the primary key table.
+  google.protobuf.StringValue schema = 2;
+
+  // Specifies the primary key table to get the foreign keys for.
+  string table = 3;
+}
+
+// SQL Execution Action Messages
+
+/*
+ * Request message for the "CreatePreparedStatement" action on a Flight SQL enabled backend.
+ */
+message ActionCreatePreparedStatementRequest {
+  option (experimental) = true;
+
+  // The valid SQL string to create a prepared statement for.
+  // The query should be treated as an opaque value, that is, clients should not attempt to parse this.
+  string query = 1;
+}
+
+/*
+ * Wrap the result of a "GetPreparedStatement" action.
+ */
+message ActionCreatePreparedStatementResult {
+  option (experimental) = true;
+
+  // Opaque handle for the prepared statement on the server.
+  bytes prepared_statement_handle = 1;
+
+  // If a result set generating query was provided, dataset_schema contains the 
+  // schema of the dataset as described in Schema.fbs::Schema, it is serialized as an IPC message.
+  bytes dataset_schema = 2;
+
+  // If the query provided contained parameters, parameter_schema contains the 
+  // schema of the expected parameters as described in Schema.fbs::Schema, it is serialized as an IPC message.
+  bytes parameter_schema = 3;
+}
+
+/*
+ * Request message for the "ClosePreparedStatement" action on a Flight SQL enabled backend.
+ * Closes server resources associated with the prepared statement handle.
+ */
+message ActionClosePreparedStatementRequest {
+  option (experimental) = true;
+
+  // Opaque handle for the prepared statement on the server.
+  bytes prepared_statement_handle = 1;
+}
+
+
+// SQL Execution Messages.
+
+/*
+ * Represents a SQL query. Used in the command member of FlightDescriptor
+ * for the following RPC calls:
+ *  - GetSchema: return the schema of the query.
+ *  - GetFlightInfo: execute the query.
+ */
+message CommandStatementQuery {
+  option (experimental) = true;
+
+  // The SQL syntax.
+  // The query should be treated as an opaque value, that is, clients should not attempt to parse this.
```

Comment:
Just curious on why the warning that clients shouldn't attempt to parse it?

Is this because the SQL query could be any SQL dialect?

Reply:
Oops, I think that was meant to go below: https://github.com/apache/arrow/pull/10906/files#r692388125

The ticket shouldn't be parsed since it's private to the server. Of course, the client can parse this field here since it's the one generating it in the first place.